// @portabletext/editor/lib/utils/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 });
var util_sliceBlocks = require("../_chunks-cjs/util.slice-blocks.cjs"), util_blockOffsetsToSelection = require("../_chunks-cjs/util.block-offsets-to-selection.cjs"), util_reverseSelection = require("../_chunks-cjs/util.reverse-selection.cjs");
function isTextBlock(context, block) {
  return block._type === context.schema.block.name;
}
function mergeTextBlocks({
  context,
  targetBlock,
  incomingBlock
}) {
  const parsedIncomingBlock = util_blockOffsetsToSelection.parseBlock({
    context,
    block: incomingBlock,
    options: {
      refreshKeys: !0
    }
  });
  return !parsedIncomingBlock || !isTextBlock(context, parsedIncomingBlock) ? targetBlock : {
    ...targetBlock,
    children: [...targetBlock.children, ...parsedIncomingBlock.children],
    markDefs: [...targetBlock.markDefs ?? [], ...parsedIncomingBlock.markDefs ?? []]
  };
}
function splitTextBlock({
  context,
  block,
  point
}) {
  const firstChild = block.children.at(0), lastChild = block.children.at(block.children.length - 1);
  if (!firstChild || !lastChild)
    return;
  const before = util_sliceBlocks.sliceBlocks({
    blocks: [block],
    selection: {
      anchor: {
        path: [{
          _key: block._key
        }, "children", {
          _key: firstChild._key
        }],
        offset: 0
      },
      focus: point
    }
  }).at(0), after = util_sliceBlocks.sliceBlocks({
    blocks: [block],
    selection: {
      anchor: point,
      focus: {
        path: [{
          _key: block._key
        }, "children", {
          _key: lastChild._key
        }],
        offset: util_sliceBlocks.isSpan(context, lastChild) ? lastChild.text.length : 0
      }
    }
  }).at(0);
  if (!(!before || !after) && !(!isTextBlock(context, before) || !isTextBlock(context, after)))
    return {
      before,
      after
    };
}
exports.blockOffsetToSpanSelectionPoint = util_sliceBlocks.blockOffsetToSpanSelectionPoint;
exports.getBlockEndPoint = util_sliceBlocks.getBlockEndPoint;
exports.getBlockStartPoint = util_sliceBlocks.getBlockStartPoint;
exports.getTextBlockText = util_sliceBlocks.getTextBlockText;
exports.isEmptyTextBlock = util_sliceBlocks.isEmptyTextBlock;
exports.isEqualSelectionPoints = util_sliceBlocks.isEqualSelectionPoints;
exports.isKeyedSegment = util_sliceBlocks.isKeyedSegment;
exports.isSpan = util_sliceBlocks.isSpan;
exports.sliceBlocks = util_sliceBlocks.sliceBlocks;
exports.spanSelectionPointToBlockOffset = util_sliceBlocks.spanSelectionPointToBlockOffset;
exports.blockOffsetsToSelection = util_blockOffsetsToSelection.blockOffsetsToSelection;
exports.reverseSelection = util_reverseSelection.reverseSelection;
exports.isTextBlock = isTextBlock;
exports.mergeTextBlocks = mergeTextBlocks;
exports.splitTextBlock = splitTextBlock;
